client.go

```diff
@@ -0,0 +1,260 @@
+// them somewhere. It maintains its own sending queue for events, distinct from
+// both the package-level libhoney queue and any other client. Clients should be
+// created with NewClient(config). A manually created Client{} will function as
+// a nil output and drop everything handed to it (so can be safely used in dev
+// and tests). For more complete testing you can create a Client with a
+// MockOutput transmission then inspect the events it would have sent.
+type Client struct {
+	transmission transmission.Sender
+	logger       Logger
+	builder      *Builder
+
+	oneTx      sync.Once
+	oneLogger  sync.Once
+	oneBuilder sync.Once
+}
+
+// ClientConfig is a subset of the global libhoney config that focuses on the
+// configuration of the client itself. The other config options are specific to
+// a given transmission Sender and should be specified there if the defaults
+// need to be overridden.
+type ClientConfig struct {
+	// APIKey is the Honeycomb authentication token. If it is specified during
+	// libhoney initialization, it will be used as the default API key for all
+	// events. If absent, API key must be explicitly set on a builder or
+	// event. Find your team's API keys at https://ui.honeycomb.io/account
+	APIKey string
+
+	// Dataset is the name of the Honeycomb dataset to which to send these events.
+	// If it is specified during libhoney initialization, it will be used as the
+	// default dataset for all events. If absent, dataset must be explicitly set
+	// on a builder or event.
+	Dataset string
+
+	// SampleRate is the rate at which to sample this event. Default is 1,
+	// meaning no sampling. If you want to send one event out of every 250 times
+	// Send() is called, you would specify 250 here.
+	SampleRate uint
+
+	// APIHost is the hostname for the Honeycomb API server to which to send this
+	// event. default: https://api.honeycomb.io/
+	APIHost string
+
+	// Transmission allows you to override what happens to events after you call
+	// Send() on them. By default, events are asynchronously sent to the
+	// Honeycomb API. You can use the MockOutput included in this package in
+	// unit tests, or use the transmission.WriterSender to write events to
+	// STDOUT or to a file when developing locally.
+	Transmission transmission.Sender
+
+	// Logger defaults to nil and the SDK is silent. If you supply a logger here
+	// (or set it to &DefaultLogger{}), some debugging output will be emitted.
+	// Intended for human consumption during development to understand what the
+	// SDK is doing and diagnose trouble emitting events.
+	Logger Logger
+}
+
+// NewClient creates a Client with defaults correctly set
+func NewClient(conf ClientConfig) (*Client, error) {
+	if conf.SampleRate == 0 {
+		conf.SampleRate = defaultSampleRate
+	}
+	if conf.APIHost == "" {
+		conf.APIHost = defaultAPIHost
+	}
+	if conf.Dataset == "" {
+		conf.Dataset = defaultDataset
+	}
+
+	c := &Client{
+		logger: conf.Logger,
+	}
+	c.ensureLogger()
```

Comment:
I’m not clear why that would be better. The way it is not the sync.Once gets triggered during initialization so all future calls quickly noop. Delaying that till the first place they’re hit makes for a more likely collision (and therefore block) doesn’t it? 

Can you say more about the motivation for this suggested change?

Reply:
It just makes initialization more consistent - right now there are 3 different patterns for `conf` fallback to initialize different fields in the `Client` struct.

The behavior you're trying to avoid (no `ensure*` call leading to `sync.Once` collision later) is also how the `c.transmission` initialization below acts :/

How about both?

1. change the initialization pattern (for everything but `.builder`, since it requires the backref) to be consistent:
  * Check `conf` setting for nil, fallback by putting the one we want in `conf`
  * `Client` init solely out of `conf`
2. post `Client` init, call c.transmission.Start() so we can early error, then call both `ensureLogger()` and `ensureTransmission()` to use up their `sync.Once`'s before returning the client.